libsupport/include/katana/ArrowVisitor.h

```diff
@@ -197,47 +198,49 @@ struct ArrowDispatcher {
+  // No further changes will be necessary to enable new types through arrow 6.0
+  // The only change required to bump arrow versions is NUM_TYPES
   template <typename... ArrowTypes, typename Visitor, typename... Args>
   static typename std::decay_t<Visitor>::ResultType Dispatch(
       Visitor&& visitor, Args&&... args) {
     constexpr auto index = sizeof...(ArrowTypes);
     auto type_id =
         GetArrowTypeID(std::get<index>(std::forward_as_tuple(args...)));
 
-    switch (type_id) {
-#define TYPE_CASE(EnumType)                                                    \
-  case arrow::Type::EnumType: {                                                \
+    // arrow::Type::type is an enum covering the universe of Arrow types
+    // This enum is guaranteed by construction to have:
+    // NA = 0 (null type), MAX_ID = <number of types>
+    // arrow 4.0 has 37 types, arrow 6.0 has 38
+#define NUM_TYPES 38
+    static_assert(arrow::Type::MAX_ID == NUM_TYPES);
+
+    // API for this macro gives (z, index, data)
+    // z is used to allow PP_REPEAT to nest arbitrarily
+    // index is distinct for each repetition: [0, NUM_REPEATS)
+    // data allows arbitrary passthrough, unchanged for each iteration
+    //   this usage does not require any passthrough data
+#define ENUM_TYPE_CASE(z, index, unused)                                       \
+  case static_cast<arrow::Type::type>(index): {                                \
     using ArrowType =                                                          \
-        typename arrow::TypeIdTraits<arrow::Type::EnumType>::Type;             \
+        typename arrow::TypeIdTraits<static_cast<arrow::Type::type>(           \
+            index)>::Type;                                                     \
     return Call<ArrowTypes..., ArrowType>(                                     \
         std::forward<Visitor>(visitor), std::forward<Args>(args)...);          \
   }
-      TYPE_CASE(INT8)
-      TYPE_CASE(UINT8)
-      TYPE_CASE(INT16)
-      TYPE_CASE(UINT16)
-      TYPE_CASE(INT32)
-      TYPE_CASE(UINT32)
-      TYPE_CASE(INT64)
-      TYPE_CASE(UINT64)
-      TYPE_CASE(FLOAT)
-      TYPE_CASE(DOUBLE)
-      TYPE_CASE(BOOL)
-      TYPE_CASE(DATE32)     // since UNIX epoch in days
-      TYPE_CASE(DATE64)     // since UNIX epoch in millis
-      TYPE_CASE(TIME32)     // since midnight in seconds or millis
-      TYPE_CASE(TIME64)     // since midnight in micros or nanos
-      TYPE_CASE(TIMESTAMP)  // since UNIX epoch in seconds or smaller
-      TYPE_CASE(STRING)     // TODO(daniel) DEPRECATED
-      TYPE_CASE(LARGE_STRING)
-      TYPE_CASE(STRUCT)
-      TYPE_CASE(LIST)  // TODO(daniel) DEPRECATED
-      TYPE_CASE(LARGE_LIST)
-      TYPE_CASE(NA)
-#undef TYPE_CASE
+
+    switch (type_id) {
+      // repeat the ENUM_TYPE_CASE macro NUM_TYPES times
+      // ~ is an unparsed token as recommended by Boost docs
+      BOOST_PP_REPEAT(NUM_TYPES, ENUM_TYPE_CASE, ~);
```

Comment:
If you really wanted to do this with template metaprogramming, you could try:

```
template<int i> Dispatch(...) {
  if (v == i) {
    ...
  } else {
    return Dispatch<i-1>(...)
  }
}
```

I don't know if this would be any easier to debug though, and you should check the microbenchmark before and after.